service/policy/db/namespaces.go

```diff
@@ -386,3 +398,212 @@ func (c PolicyDBClient) RemovePublicKeyFromNamespace(ctx context.Context, k *nam
+	}
+
+	// Return the full certificate object
+	return c.GetCertificate(ctx, certID)
+}
+
+// GetCertificate retrieves a certificate by its ID
+func (c PolicyDBClient) GetCertificate(ctx context.Context, id string) (*policy.Certificate, error) {
+	cert, err := c.queries.getCertificate(ctx, id)
+	if err != nil {
+		return nil, db.WrapIfKnownInvalidQueryErr(err)
+	}
+
+	metadata := &common.Metadata{}
+	if err = unmarshalMetadata(cert.Metadata, metadata); err != nil {
+		return nil, err
+	}
+
+	return &policy.Certificate{
+		Id:       cert.ID,
+		Pem:      cert.Pem,
+		Metadata: metadata,
+	}, nil
+}
+
+// DeleteCertificate removes a certificate from the database
+func (c PolicyDBClient) DeleteCertificate(ctx context.Context, id string) error {
+	count, err := c.queries.deleteCertificate(ctx, id)
+	if err != nil {
+		return db.WrapIfKnownInvalidQueryErr(err)
+	}
+	if count == 0 {
+		return db.ErrNotFound
+	}
+	return nil
+}
+
+// resolveNamespaceID resolves a namespace identifier to its UUID
+func (c PolicyDBClient) resolveNamespaceID(ctx context.Context, identifier *common.IdFqnIdentifier) (string, error) {
+	// If ID is provided, check if it's a valid UUID
+	if identifier.GetId() != "" {
+		id := identifier.GetId()
+		// Check if the ID is a valid UUID
+		uuid := pgtypeUUID(id)
+		if uuid.Valid {
+			// It's a valid UUID, use it directly
+			return id, nil
+		}
+		// Not a valid UUID, treat it as a namespace name and look it up
+		ns, err := c.GetNamespace(ctx, &namespaces.GetNamespaceRequest_Fqn{Fqn: id})
+		if err != nil {
+			return "", err
+		}
+		return ns.GetId(), nil
+	}
+
+	// If FQN is provided, look up the namespace by FQN to get its ID
+	if identifier.GetFqn() != "" {
+		ns, err := c.GetNamespace(ctx, &namespaces.GetNamespaceRequest_Fqn{Fqn: identifier.GetFqn()})
+		if err != nil {
+			return "", err
+		}
+		return ns.GetId(), nil
+	}
+	return "", errors.Join(db.ErrUnknownSelectIdentifier, fmt.Errorf("type [%T] value [%v]", identifier, identifier))
+}
+
+// AssignCertificateToNamespace assigns a trusted root certificate to a namespace for trust validation
+func (c PolicyDBClient) AssignCertificateToNamespace(ctx context.Context, namespaceIdentifier *common.IdFqnIdentifier, certificateID string) error {
+	namespaceID, err := c.resolveNamespaceID(ctx, namespaceIdentifier)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.queries.assignCertificateToNamespace(ctx, assignCertificateToNamespaceParams{
+		NamespaceID:   namespaceID,
+		CertificateID: certificateID,
+	})
+	if err != nil {
+		return db.WrapIfKnownInvalidQueryErr(err)
+	}
+	return nil
+}
+
+// CreateAndAssignCertificateToNamespace creates a certificate and assigns it to a namespace in a transaction
+func (c PolicyDBClient) CreateAndAssignCertificateToNamespace(ctx context.Context, namespaceID *common.IdFqnIdentifier, pem string, metadata []byte) (string, error) {
+	// Check if certificate with same PEM already exists
+	existingCert, err := c.queries.getCertificateByPEM(ctx, pem)
```

Comment:
There is a check then act race condition here.  The read happening out of the transaction means that multiple concurrent clients could see no certificate and fall to the transaction below, resulting in multiple certificates being generated.

Reply:
fixed